[firebase-analytics] Add native Firebase Analytics

IjzerenHein opened the pull request:
# Why

Firebase Analytics is currently not support on the Expo client because is requires certain browser features (it dynamically loads gtag.js, which in turn also requires browser features). You can install native Firebase Analytics on bare-projects, but there is no solution for the managed workflow.

# How

This PR add support for native Firebase Analytics (written from scratch by @EvanBacon) to the managed workflow.  This is fully support on standalone builds and custom clients. On the standard Expo client there is a limitation which prevents loaded projects from using native analytics (FirebaseAnalytics in a singleton and cannot be succesfuly updated after initial configuration). For that case, a dummy shim is used which logs the events to the console (for debugging purposes) & generates a one-time warning.
A pure JS based analytics tracking is currently being researched as a workaround solution.

# Test Plan

- Tests (Firebase Analytics) have been added to the test-suite
- You can run the tests on bare and should see analytics in the Expo Custom Client Firebase project

## Running on Expo client

The tests in the test-suite are conditional and will run different checks whether a `googleServicesFile` is found in app.json or not. By default the test-suite does not (yet) have a `googleServicesFile` config for iOS or Android. This means that the standard client will run tests to verify that the analytics functions fail (because of the missing config).

To run the the other tests, add `googleServicesFile` configs to `app.json` of test-suite. 

```json
{
  "expo": {
    "ios": {
      "bundleIdentifier": "io.expo.testsuite",
      "googleServicesFile": "./GoogleService-Info.plist"
     },
     "android": {
      "googleServicesFile": "./google-services.json"
     }
  }
}
```

Now when running the tests in the expo client, it will validate that the Analytics functions can be called correctly. Because of the previously mentioned limitation, it will now log the analytics events to the console.

IjzerenHein committed:
[firebase-core] Add tests to test-suite

IjzerenHein committed:
[fireabse-analytics] Add expo-firebase-analytics package

IjzerenHein committed:
[firebase-analytics] Add tests to test-suite

IjzerenHein committed:
[bare-expo] Update iOS app-icon assets

IjzerenHein committed:
[firebase-analytics] Add expo-firebase-analytics to Expo client

IjzerenHein committed:
[bare-expo] Add firebase-analytics to BasePackageList.java

IjzerenHein committed:
[firebase-analytics] Add firebase-analytics to bundleNativeModules.json

IjzerenHein committed:
[docs] Add firebase-analytics docs

IjzerenHein committed:
[firebase-core] Add firebase-core to Expo client iOS project file

IjzerenHein committed:
[firebase-analytics] Update yarn.lock after adding `firebase` (js sdk) dependency

IjzerenHein committed:
[firebase-analytics] Add js build output

IjzerenHein committed:
[firebase-analytics] Update sdk docs

IjzerenHein committed:
[firebase-analytics] Update SDK docs

IjzerenHein committed:
[firebase-analytics] Remove firebase JS dep to check CI

IjzerenHein committed:
[docs] Update SDK docs & readme for expo-firebase-analytics

IjzerenHein committed:
[expo-analytics] Remove firebase peer-dependency

IjzerenHein committed:
[firebase-analytics] Add web-support & code cleanup

IjzerenHein committed:
[docs] Update firebase-analytics docs

IjzerenHein committed:
[firebase-analytics] Update build files

IjzerenHein committed:
[firebase-analytics] Fix outdated build files

IjzerenHein committed:
[firebase-analytics] Fix lint warnings

brentvatne commented on packages/expo-firebase-analytics/android/src/main/java/expo/modules/firebase/analytics/FirebaseAnalyticsModule.java:
```diff
@@ -0,0 +1,176 @@
+    try {
+      systemApp = FirebaseApp.getInstance();
+    } catch (Exception e) {
+      // nop
```
maybe we should log this, it looks like it will throw an `IllegalStateException` if the app isn't initialized yet

IjzerenHein replied:
This is just for checking whether the default firebase app is initialized. Unfortunately that's not possible without an exception handler here. If no default app exists, then that's not an error situation per-se.

brentvatne commented on packages/expo-firebase-analytics/android/src/main/java/expo/modules/firebase/analytics/FirebaseAnalyticsModule.java:
```diff
@@ -0,0 +1,176 @@
+
+  @ExpoMethod
+  public void setCurrentScreen(final String screenName, final String screenClassOverride, final Promise promise) {
+    // This is the only method that runs on the main thread.
```
interesting.. I wonder why this is required, it seems like the Firebase docs don't really explain it

IjzerenHein replied:
True, it just says: "This method must be called from the app's main thread.".

brentvatne reviewed (commented):
just a couple comments, this is looking good!

EvanBacon commented on packages/expo-firebase-analytics/README.md:
```diff
@@ -0,0 +1,38 @@
+
+# Setup native Firebase
+
+Follow the [Setup Native Firebase guide](https://github.com/expo/expo/blob/master/docs/pages/versions/unversioned/guides/setup-native-firebase.md) on how to setup the native Firebase SDK & configuration further.
```
reads a little hard
```suggestion
Follow the ["Setup Native Firebase"](https://github.com/expo/expo/blob/master/docs/pages/versions/unversioned/guides/setup-native-firebase.md) guide in the docs to get started using native Firebase SDK features.
```

EvanBacon commented on packages/expo-firebase-analytics/android/build.gradle:
```diff
@@ -0,0 +1,76 @@
+apply plugin: 'maven'
+
+group = 'host.exp.exponent'
+version = '3.0.0'
```
```suggestion
version = '1.0.0'
```

EvanBacon commented on packages/expo-firebase-analytics/android/build.gradle:
```diff
@@ -0,0 +1,76 @@
+  defaultConfig {
+    minSdkVersion safeExtGet("minSdkVersion", 21)
+    targetSdkVersion safeExtGet("targetSdkVersion", 28)
+    versionCode 4
```
```suggestion
    versionCode 1
```

EvanBacon replied:
I remember why it's like this now, nevermind 😅

EvanBacon commented on packages/expo-firebase-analytics/android/build.gradle:
```diff
@@ -0,0 +1,76 @@
+    minSdkVersion safeExtGet("minSdkVersion", 21)
+    targetSdkVersion safeExtGet("targetSdkVersion", 28)
+    versionCode 4
+    versionName '3.0.0'
```
```suggestion
    versionName '1.0.0'
```

EvanBacon commented on packages/expo-firebase-analytics/ios/EXFirebaseAnalytics/EXFirebaseAnalytics.m:
```diff
@@ -0,0 +1,153 @@
+    reject(@"ERR_FIREBASE_ANALYTICS", @"EXFirebaseCore could not be found. Ensure that your app has correctly linked 'expo-firebase-core' and your project has react-native-unimodules installed.", nil);
+    return nil;
+  }
+  FIRApp* defaultApp = [_firebaseCore defaultApp];
```
```suggestion
  FIRApp *defaultApp = [_firebaseCore defaultApp];
```

IjzerenHein replied:
Thanks, yep made that mistake everywhere. Fixed it 👍

EvanBacon commented on packages/expo-firebase-analytics/package.json:
```diff
@@ -0,0 +1,45 @@
+  },
+  "unimodulePeerDependencies": {
+    "@unimodules/core": "*"
+  },
```
move this to `peerDependencies ` and delete `unimodulePeerDependencies `

EvanBacon commented on packages/expo-firebase-analytics/src/Analytics.ts:
```diff
@@ -0,0 +1,177 @@
+  if (!ExpoFirebaseAnalytics.setUnavailabilityLogging) {
+    throw new UnavailabilityError('expo-firebase-analytics', 'setUnavailabilityLogging');
+  }
+  return ExpoFirebaseAnalytics.setUnavailabilityLogging(isEnabled);
```
```suggestion
  ExpoFirebaseAnalytics.setUnavailabilityLogging(isEnabled);
```
or set the return type to `Promise<void>`. All unimodule methods are currently async.

EvanBacon commented on packages/expo-firebase-analytics/src/ExpoFirebaseAnalytics.ts:
```diff
@@ -0,0 +1,78 @@
+    if (isUnavailabilityLoggingEnabled) {
+      if (!isUnavailabilityWarningLogged) {
+        console.warn(
+          'Firebase Analytics is not available in the Expo client. To test Firebase Analytics create a stand-alone build or custom client. To suppress this warning use `setUnavailabilityLogging(false)`.'
```
```suggestion
          'Firebase Analytics is not available in the Expo client. To test Firebase Analytics create a standalone build or custom client. To suppress this warning use `setUnavailabilityLogging(false)`.'
```

EvanBacon reviewed (commented):
Nice work

IjzerenHein committed:
[firebase-analytics] Fixed code-style errors & typos after review

IjzerenHein committed:
[firebase-analytics] Moved `unimodulePeerDependencies` to regular peerDependencies in package.json

IjzerenHein committed:
[firebase-analytics] Update js build output

IjzerenHein commented:
> Nice work

Thanks 😄

IjzerenHein committed:
[fireabse-analytics] Update changelog

IjzerenHein committed:
[firebse-analytics] Update Android build.gradle

IjzerenHein committed:
[ios] Update pods

IjzerenHein committed:
[firebase-analytics] Add recording of `screen_view` event when calling `setCurrentScreen`

On web, calling `setCurrentScreen` does not automatically record a screen_view event as
it does on native. We therefore record the 'screen_view' event manually.
https://stackoverflow.com/questions/59330467/how-to-track-page-view-with-firebase-analytics-in-a-web-single-page-app

IjzerenHein committed:
[firebase-analytics] Update JS build files

brentvatne commented on packages/expo-firebase-analytics/package.json:
```diff
@@ -35,10 +35,8 @@
-    "@unimodules/core": "*"
-  },
   "peerDependencies": {
+    "@unimodules/core": "*",
```
the reason we don't put this as a regular peer dep is that if you install `react-native-unimodules` then `@unimodules/core` is a dep of that, and peer deps can only be satisfied by direct dependencies so even though the code will work the user will get a warning on install

brentvatne replied:
```suggestion
```

IjzerenHein committed:
[firebase-analytics] Remove @unimodules/core peerDependency from package.json

alexbchr commented:
I have 2 questions related to this PR:
1. Will this be included in Expo 37? I don't see it planned in the roadmap.
2. Can we expect other native Firebase SDK integrations planned soon? I am mostly interested by Firestore and Dynamic Links integrations.

IjzerenHein commented:
@alexbchr Yes this is planned for Expo 37 👍 

As for other integrations, I cannot give you any assurances, but yet we are working on bringing more Firebase services to Expo.

Dynamic Links will likely not be possible in the Expo client though, as it requires an URL scheme linked into the app. 

Regarding Firestore, you can currently use it with the Firebase JS SDK. Does this provide any limitations for you?

alexbchr commented:
> @alexbchr Yes this is planned for Expo 37 

Thanks a lot for the info! I am really looking forward to this!

> As for other integrations, I cannot give you any assurances, but yet we are working on bringing more Firebase services to Expo.
> 
> Dynamic Links will likely not be possible in the Expo client though, as it requires an URL scheme linked into the app.
> 
> Regarding Firestore, you can currently use it with the Firebase JS SDK. Does this provide any limitations for you?

Good to know! And yes for Dynamic Links I understand the limitations. It would however be great if we can use them in standalone apps! As to Firestore, I currently use the Firebase JS SDK and it works quite well, but I encountered some performance issues with it, as it runs on the JS thread. I think a native integration would hopefully be a little bit more performant with fast successive updates.

cemo commented:
@IjzerenHein Will you add other firebase features support as well? 😇

IjzerenHein commented:
> @IjzerenHein Will you add other firebase features support as well? 😇

Yes, that is definitely our goal.

Analytics is ready we will publish an update for Phone auth soon as well.

I think this sums up the situation and goals for Firebase nicely 👍 

> One of the problems is that some parts of the native Firebase SDK are written for single app environments and are tightly coupled with other services. This makes running the native Firebase SDK in the multi-app Expo client a challenge and in some cases, like Analytics, even impossible. Other Firebase services, like Dynamic Links and some Auth providers, require that URL schemes are defined in the app, otherwise they won't work.
>
> Lastly, we need to minimize dependencies on external services in packages included in the Expo client that we ship to the stores. The core of the Expo SDK will be free of third party service dependencies in the long-run, and those libraries will instead be available to developers through our upcoming build service. This is still evolving and we will speak about it more in the future.
>
> We are looking forward to bringing as much of the Firebase functionality to our users as possible. If you have Firebase related feature requests, feel free to add or upvote them in our [feature request tracker](https://expo.canny.io/).
